src/client/dfuse/ops/forget.c

```diff
@@ -8,58 +8,51 @@
 #include "dfuse.h"
 
 static void
-dfuse_forget_one(struct dfuse_projection_info *fs_handle,
-		 fuse_ino_t ino, uintptr_t nlookup)
+dfuse_forget_one(struct dfuse_info *dfuse_info, fuse_ino_t ino, uintptr_t nlookup)
 {
-	d_list_t *rlink;
-	int rc;
+	struct dfuse_inode_entry *ie;
+	int                       rc;
 
-	/* One additional reference is needed because the rec_find() itself
-	 * acquires one
-	 */
+	/* One additional reference is needed because the rec_find() itself acquires one */
 	nlookup++;
 
-	rlink = d_hash_rec_find(&fs_handle->dpi_iet, &ino, sizeof(ino));
-	if (!rlink) {
-		DFUSE_TRA_WARNING(fs_handle, "Unable to find ref for %#lx %lu",
-				  ino, nlookup);
+	ie = dfuse_inode_lookup(dfuse_info, ino);
+	if (!ie) {
+		DFUSE_TRA_WARNING(dfuse_info, "Unable to find ref for %#lx %lu", ino, nlookup);
 		return;
 	}
 
-	DFUSE_TRA_DEBUG(container_of(rlink, struct dfuse_inode_entry, ie_htl),
-			"inode %#lx count %lu",
-			ino, nlookup);
+	DFUSE_TRA_DEBUG(ie, "inode %#lx count %lu", ino, nlookup);
 
-	rc = d_hash_rec_ndecref(&fs_handle->dpi_iet, nlookup, rlink);
-	if (rc != -DER_SUCCESS) {
-		DFUSE_TRA_ERROR(fs_handle, "Invalid refcount %lu on %p",
-				nlookup,
-				container_of(rlink, struct dfuse_inode_entry,
-					     ie_htl));
-	}
+	rc = d_hash_rec_ndecref(&dfuse_info->dpi_iet, nlookup, &ie->ie_htl);
+	if (rc != -DER_SUCCESS)
+		DFUSE_TRA_ERROR(dfuse_info, "Invalid refcount %lu on %p", nlookup, ie);
 }
 
 void
 dfuse_cb_forget(fuse_req_t req, fuse_ino_t ino, uintptr_t nlookup)
 {
-	struct dfuse_projection_info *fs_handle = fuse_req_userdata(req);
+	struct dfuse_info *dfuse_info = fuse_req_userdata(req);
 
 	fuse_reply_none(req);
 
-	dfuse_forget_one(fs_handle, ino, nlookup);
+	dfuse_forget_one(dfuse_info, ino, nlookup);
 }
 
 void
-dfuse_cb_forget_multi(fuse_req_t req, size_t count,
-		      struct fuse_forget_data *forgets)
+dfuse_cb_forget_multi(fuse_req_t req, size_t count, struct fuse_forget_data *forgets)
 {
-	struct dfuse_projection_info *fs_handle = fuse_req_userdata(req);
-	int i;
+	struct dfuse_info *dfuse_info = fuse_req_userdata(req);
+	int                i;
 
 	fuse_reply_none(req);
 
-	DFUSE_TRA_DEBUG(fs_handle, "Forgetting %zi", count);
+	DFUSE_TRA_DEBUG(dfuse_info, "Forgetting %zi", count);
+
+	D_RWLOCK_RDLOCK(&dfuse_info->di_forget_lock);
```

Comment:
D_RWLOCK_RDLOCK, not D_RWLOCK_WRLOCK?
Is this because d_hash_rec_ndecref has its own lock?

Reply:
I want forgets to happen in parallel as they do now so forget takes a reader lock (although it does modify - this is protected by the hash table) but I also want the ioctl to block whilst there are pending forget operations so that takes a writer lock.  It's the only way I could think of for the ioctl to pause which pending operations are happening.  The kernel seems to send a small number of forgetmany operation, each with a large number of forgets so this won't guarantee that the kernel has completed but it does at least pause the reader for any in-flight to complete.